### ISSUE TITLE
valid UTF-8 causing invalid UTF-8 error on update

### ISSUE THREAD

**Issue by rahaeli** [dreamwidth/dw-free#1331]
http://www.dreamwidth.org/support/see_request?id=30348

User provides text in the support request including single curly-quotes ( ‘Decadent Sisters’ ) and ellipsis ( A new journey begins… ) that, when posted to DW using either the old update page or the new update page, causes the "invalid UTF-8" error. Removing those three characters makes it work.

I don't think it's actually invalid UTF-8, so ... I got nothin'.

**Comment by woggy** [comment]
claimed.

**Comment by woggy** [comment]
Notes and observations: in addition to the curlyquotes and ellipsis noted by rah, there is *also* an suspected-unintentional character (Unicode codepoint 0x0008, "backspace") before the "2000" text in OP's text string.

Backspace + curlyquotes/ellipsis: Fails
Backspace alone: Passes
curlyquotes/ellipsis alone: Passes

Attempting to trace the validater back to handle it directly and see what's going on, and I've ended up at cgi-bin/LJ/TextUtil.pm line 338, which is outsourcing the check to something labelled Unicode::CheckUTF8...which is only elsewise referenced in association with an empty pair of brackets in checkconfig.pl.

Am wanting to run said external validater directly to see whether it will properly parse the full set of text and possibly we're mangling it on the way to the passdown, but remain stumped by not being able to see where it even resides, much less the proper ways to call it.

(as a side note, continue to confirm that LJ's validater does not throw errors at the combination, but this is of limited use given that aiui cannot directly compare code anymore.)

**Comment by rahaeli** [comment]
It's a CPAN module:

   http://search.cpan.org/~bradfitz/Unicode-CheckUTF8-1.03/CheckUTF8.pm

**Comment by woggy** [comment]
Right okay so. Issue in this case is resolved by stripping out the low-ASCII control character that shouldn't have been there in the first place. Working on a new function-call to programmatically strip nonprinting low-ASCII from entries (and comments?) before passing to the text_in() call.

**Comment by zorkian** [comment]
I don't think that's the right approach? Because sometimes those characters are part of UTF8 sequences aren't they?

I mean, the validator did the correct thing here. The user gave it shit input, it said the input was shit. I mean, we could possibly be clearer "if you copied this text from some other text, please try round-tripping through Notepad to clear special formatting"?

**Comment by rahaeli** [comment]
The validator definitely did the correct thing! However, us stripping out things we know the validator will choke on (and that have limited/no valid use case in our particular environment) before giving it to the validator, is an opportunity to improve user experience. The invisible control characters the validator chokes on are really hard to identify and remove on the user end -- I actually don't think Notepad even clears that one -- and quietly correcting the user's input to something the validator doesn't choke on saves the user a ton of frustration. Better to Do The Right Thing on our end and save the user time trying to debug.

I think the use case "accidentally inserted an invisible control character without realizing what they were doing" is way more frequent than "user deliberately needs to insert that Unicode backspace character".

**Comment by zorkian** [comment]
That's not my point.

You can't arbitrarily strip low-ASCII characters because it makes something pass. That's akin to pulling wires from a circuitboard until you "get it to turn on". I mean, it might work, and you definitely will find cases where it does, but you'll also find cases where doing so removes something necessary.

The only valid way of wording this argument is: "I've read the UTF-8 spec, and because of reasons X, Y, Z, it's safe to remove characters in range \xNN-\xMM at the beginning of sequences." Or, something to that effect.

If you base the patch around the spec and can justify that it actually works as expected in all cases, then by all means let's modify user input. But otherwise, it's not safe.

**Comment by zorkian** [comment]
After more investigation with @woggy, it turns out the validator is actually not doing the right thing at all!!!!!! Wat.

In more time and investigation, we think that we should rip out `Unicode::CheckUTF8` and use something like:

```
#!/usr/bin/perl

our $valid_utf8_regexp = qr/
        [\x{00}-\x{7f}]
      | [\x{c2}-\x{df}][\x{80}-\x{bf}]
      |         \x{e0} [\x{a0}-\x{bf}][\x{80}-\x{bf}]
      | [\x{e1}-\x{ec}][\x{80}-\x{bf}][\x{80}-\x{bf}]
      |         \x{ed} [\x{80}-\x{9f}][\x{80}-\x{bf}]
      | [\x{ee}-\x{ef}][\x{80}-\x{bf}][\x{80}-\x{bf}]
      |         \x{f0} [\x{90}-\x{bf}][\x{80}-\x{bf}]
      | [\x{f1}-\x{f3}][\x{80}-\x{bf}][\x{80}-\x{bf}][\x{80}-\x{bf}]
      |         \x{f4} [\x{80}-\x{8f}][\x{80}-\x{bf}][\x{80}-\x{bf}]
/x;

print "about \xf1\x08 2000" =~ /^$valid_utf8_regexp*$/g ? "true" : "false";
print "\n";
```

**Comment by afuna** [comment]
noideapuppy thumbsup


On Thu, Jun 25, 2015, at 02:19 PM, Mark Smith wrote:
> After more investigation with @woggy[1], it turns out the validator is
> actually not doing the right thing at all!!!!!! Wat.


> In more time and investigation, we think that we should rip out
> Unicode::CheckUTF8 and use something like:


> #!/usr/bin/perl

our $valid_utf8_regexp = qr/        [\x{00}-\x{7f}]
      | [\x{c2}-\x{df}][\x{80}-\x{bf}]        \x{e0} [\x{a0}-\x{bf}][\x{80}-
      | \x{bf}] [\x{e1}-\x{ec}][\x{80}-\x{bf}][\x{80}-\x{bf}]
      | \x{ed} [\x{80}-\x{9f}][\x{80}-\x{bf}] [\x{ee}-\x{ef}][\x{80}-\x{bf}][\x{80}-
      | \x{bf}]        \x{f0} [\x{90}-\x{bf}][\x{80}-\x{bf}] [\x{f1}-\x{f3}][\x{80}-\x{bf}][\x{80}-\x{bf}][\x{80}-
      | \x{bf}]        \x{f4} [\x{80}-\x{8f}][\x{80}-\x{bf}][\x{80}-
      | \x{bf}]
/x;

print "about \xf1\x08 2000" =~ /^$valid_utf8_regexp*$/g ? "true" :
"false"; print "\n";
>
> — Reply to this email directly or view it on GitHub[2].


>



Links:

  1. https://github.com/woggy
  2. https://github.com/dreamwidth/dw-free/issues/1331#issuecomment-115401282

**Comment by anall** [comment]
@zorkian Unicode::CheckUTF8 claims "This is an XS wrapper around some Unicode Consortium code to check if a string is valid UTF-8, revised to conform to what expat/Mozilla think is valid UTF-8, especially with regard to low-ASCII characters." -- is that claim no longer valid or necessary?

**Comment by zorkian** [comment]
I tried to find the original material that caused that but couldn't find anybody who was saying anything about it or why, and there's no link in the readme so...

Personally, I'd keep it as is and not allow low ASCII in posts because no good can come of that. But given the choices:

a) Tell the users they have malformed input (that they might possibly not be able to see)
b) Arbitrarily delete low ASCII from their input
c) Allow all valid UTF-8 (which includes low ASCII)

It seems like C > A > B. I would prefer A. But others prefer C. And honestly, my argument against B is because it might lead to insanity where the input is valid but not what the user expected. So I guess C is probably not the worst thing.

**Comment by anall** [comment]
I do honestly wonder how *big* of an issue this exact thing is, random low-ASCII showing up in a post -- is this the only case we've seen?  Most of the time telling them to replace smart quotes helps, IIRC.

**Comment by rahaeli** [comment]
I can remember about a dozen instances of "this isn't working and we can't figure out why, even running it through Notepad to zap invisibles" in the past few years.

**Comment by zorkian** [comment]
Is that worth allowing people to post things like,

```rm -rf /^H^H^H^H^H^H^H^H# paste this code in your terminal```

In some browsers that renders as "# paste this code" but when you copy it you get the `rm -rf /`. At least, this is what I'm told. I don't really quite believe that this is valid behavior of any modern browser, but I do believe it was at some point possible.

**Comment by anall** [comment]
To be fair, my minimal testing shows:
 * Can edit to text, quotes-and-ellipsis stripped but backspace intact.
 * Can edit to text, backspace stripped but Q-E intact.
 * Can not edit with both intact.
 * Can not post with both intact.
 * Can not post with backspace intact.
 * Can post with Q-E intact, but backspace stripped.

I did a quick test with edit ( which doesn't seem to run the check ) and the full low ASCII, and it does not seem to break Chrome Linux ( besides some copypasta hell, but I am thinking that's my end, and that's attempting to post the post )

**Comment by Sophira** [comment]
For the record, the only time you'll ever see low control characters in UTF-8 is when they're representing ASCII control codes. You will never come across any byte between `\x00-\x7F` inclusive as part of a UTF-8 byte sequence that's two bytes or more - this is guaranteed by the spec. (The spec also guarantees that any byte between `\xC0-\xFF` is the first byte of a multi-byte UTF-8 sequence, and that any byte between `\x80-\xB9` is some other byte in a multi-byte UTF-8 sequence.)

That doesn't mean bytes between `\x00-\x7F` aren't valid UTF-8 characters, because they are - all valid ASCII is valid in UTF-8. In particular, you'd want to allow *some* control characters, including `\x09` (horizontal tab), `\x0A` (LF, linefeed) and perhaps `\x0D` (CR, carriage return). But stripping other control characters is something that would not affect how UTF-8 works at all.

I'm not sure it's the right solution, mind you - if control characters exist then it's probably something *else* going wrong, and stripping them might merely be masking that problem.

**Comment by woggy** [comment]
Revisiting this (because I think it's worth fixing, given what @rahaeli says about spr0t).  However, it's been a while since I looked at it - what I'm getting from the discussion so far is that we have two major options:

1: Do what the site does now - splat on invisible control characters. Possibly update the error message to indicate the possible presence thereof.

2: Change the validation routine to allow all low-ASCII.

I am of two minds about it, because on the one hand, legacy code from Brad may no longer apply - but on the other, presumably the existing validator was put in place for Good Reason.

**Comment by zorkian** [comment]
I still think A is the correct approach, but if people really feel like it's just not acceptable hen I'd be willing to go with C.

**Comment by Sophira** [comment]
I personally am a fan of B from @zorkian's choices - arbitrarily removing all low control characters other than `\x09` (tab), `\x0A` (LF) and `\x0D` (CR).

As has been noted, the other control characters are invisible, and the user may very well not be aware that they exist. There's also absolutely no reason why they should be allowed as input - they're not used to affect the browser rendering at all, Unicode has other codepoints for that. It's also going to be absolutely safe to strip these low characters even in a UTF-8-encoded string (or even stripping low bytes in a UTF-8 bytestream, thanks to the situation I described in my previous comment).

The only way it *wouldn't* be safe would be if the code assumed that the user's browser would submit in one encoding (for example, if there was a hardcoded assumption of UTF-8 input) and then the user's browser decided to submit in an encoding that didn't encode ASCII characters the same way that ASCII does - for example, multi-byte encodings like UTF-16 and ShiftJIS. That can be mitigated by making sure we check the Content-Type header of the POST request to see what encoding the browser is submitting in, and converting if necessary. (Which maybe the code already does? I'm not sure, I haven't checked.)

At this point, there's only one more situation that we haven't specifically handled - if the POST request *says* it's submitting in one encoding but submitted in another encoding anyway with incompatible characters. (For example, if somebody was trying to crack us, or if there was a major bug in a browser affecting virtually the whole Internet.) Fortunately, we get the solution to this for free - in the case of it saying it's being submitted as UTF-8 or another multi-byte encoding, it would most likely get picked up as an error when we tried to decode it, and *that* error can be relayed to the client. In any other case... well, they're probably trying to crack us, so we probably shouldn't be storing the lower bytes in the database anyway.

tl;dr: Removing the invalid control characters just makes for a better experience and doesn't cause the user to have to scour their text for the invisible control character that they didn't know was there, and it's safer than just allowing the control characters.

**Comment by rahaeli** [comment]
My preference is also still for B (stripping all low ASCII) since I haven't heard any argument for why we should accept it, and erroring out on things that users can't see and therefore can't easily correct and are accidental input 99.999% of the time is bad user experience.

**Comment by Sophira** [comment]
I think I still prefer B as I'm not sure if the user gains much from knowing where the invalid character is, but I would also be okay with A if that information was given.

Maybe a compromise? Strip as in B, but if anything was stripped then show a warning on the following page saying something like this?

> Some control characters in your submission were found and have been stripped. These characters and their locations are noted below. You do not need to resubmit; this is merely for your information.
> 
> ...ipsum dolor sit amet, consectetur**[^H]** adipiscing elit, sed do eiusmod tempor incididunt...

I'm still not entirely sure that this would help much, but it's a possibility.

**Comment by azurelunatic** [comment]
I like that! Especially since that would provide a warning to them that the
text was not as uncomplicated as they might have thought.

**Comment by woggy** [comment]
I like that as well - stripping them does seem like it'd be the Right Thing most of the time, but it does make sense to have a message about What's Going On.

**Comment by zorkian** [comment]
Okay after sitting down with the UTF-8 spec for a while, I believe this is safe:

1) Use the regexp I posted above as our new "check if valid UTF8" code,
2) If so, it is then safe to delete the low ASCII bytes we don't want.

Basically if and only if it's valid UTF-8 then I feel like we can be safe to delete the bytes that would otherwise cause problems. Does this seem reasonable? This should fix the issue to allow users to post things that are valid and we will fix the input so as not to propagate the unnecessary bytes.

**Comment by woggy** [comment]
Seems very reasonable to me.

**Comment by zorkian** [comment]
Awesome. And thanks to @Sophira for doing the major leg work to do the initial research to ensure that this path would be safe. I am pretty surprised that you can arbitrarily delete low ASCII from a UTF-8 stream and it remains valid, but it turns out the authors were pretty clever.

It would be nice to strip the same bytes that the old/current validator says are invalid. That gets us the same net result: we don't insert these bytes in our database, but it's a better user experience.

Thanks everybody for bearing with me through this. Sometimes finding the right solution takes a lot of work.

**Comment by Sophira** [comment]
My knowledge of this stuff actually comes from an earlier job doing text processing - as a result, I'm pretty knowledgable in character encoding and how to do them in Perl.

Ideally, we'd actually want to be using Perl's Encode module (which is supplied with Perl) for dealing with character sets rather than using Unicode::CheckUTF8 and other methods of converting encodings, because Perl itself and Encode has *excellent* native support for encodings and UTF-8; Encode makes it really, *really* easy to do this stuff. You can literally write a program that converts a file from one encoding to another in six short lines:

````perl
#!/usr/bin/perl

my ($in, $inenc, $out, $outenc) = @ARGV;
open(IN, "<:raw:encoding($inenc)", $in);
open(OUT, ">:raw:encoding($outenc)", $out);
while (<IN>) { print OUT; }
close(OUT);
close(IN);
````

That's literally all you need to do. It's amazing. Note that you don't even have to "use Encode;" for this - it Just Works.

(Perl programmers who don't understand this tend to be tripped up by the way Perl scalars have a UTF-8 bit which tells Perl whether the scalar should be treated as a UTF-8 string or not, and curse Perl out for it, preferring to handle scalars all in one way. This is a shame, because by doing that they miss out on a lot of the reasons why Perl handles encodings so well.)

Rewriting the code to use Encode instead of the Unicode::* modules has been something I've been wanting to do for a while, but it's too big of a task to do in an issue like this. Might be something worth bearing in mind, though.

And yes, UTF-8 is one of the cleverest schemes I've seen. I'm a big fan of it.
